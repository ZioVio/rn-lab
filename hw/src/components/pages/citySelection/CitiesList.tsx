import React from 'react';
import { FlatList, StyleSheet, View, TouchableOpacity } from 'react-native';
import ShortCity from '../../../models/shortCity';
import { AppText } from '../../shared/Text';
import City from './City';

interface CitiesListProps {
  cities: ShortCity[];
  onCityPressed: (city: ShortCity) => void;
}

export default function CitiesList({ cities, onCityPressed }: CitiesListProps) {
  const renderItem = ({ item }: { item: ShortCity }) =>
    <TouchableOpacity onPress={() => onCityPressed(item)}>
      <City city={item} />
    </TouchableOpacity>;

  const keyExtractor = (c: ShortCity) => c.id;
  return (
    <View style={styles.list}>
      {cities.length
        ? <FlatList data={cities} keyExtractor={keyExtractor} renderItem={renderItem} />
        : <NotFoundText />}
    </View>
  );
}

function NotFoundText() {
  return (
    <View style={styles.error}>
      <AppText>No cities found</AppText>
    </View>
  );
}

const styles = StyleSheet.create({
  error: {
    flexDirection: 'row',
    justifyContent: 'center',
    padding: 50,
  },
  list: {
    marginBottom: 60,
  },
});
