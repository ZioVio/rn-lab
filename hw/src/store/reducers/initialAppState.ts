import AppState from '../../models/store/appState';
import { initialState as initialAuthState } from './authReducer';
import { initialState as initialCitySelectionState } from './citySelectionReducer';
import { initialState as initialDaySelectionState } from './daySelectionReducer';
import { initialState as initialSettingsState } from './settingsReducer';
import { initialState as initialSharedState } from './sharedReducer';
import { initialState as initialWeatherDetails } from './weatherDetailsReducer';
import { initialState as initialWeatherSearchState } from './weatherSearchReducer';

const initialAppState: AppState = {
  auth: initialAuthState,
  citySelection: initialCitySelectionState,
  daySelection: initialDaySelectionState,
  settings: initialSettingsState,
  shared: initialSharedState,
  weatherDetails: initialWeatherDetails,
  weatherSearch: initialWeatherSearchState,
};

export default initialAppState;
