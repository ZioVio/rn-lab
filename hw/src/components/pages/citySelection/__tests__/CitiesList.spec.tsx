import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import CitiesList from '../CitiesList';
import ShortCity from '../../../../models/shortCity';
import WeatherState from '../../../../enums/weatherState';

import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import initialAppState from '../../../../store/reducers/initialAppState';

const mockStore = configureStore();
const createMockedStore = () => mockStore(initialAppState);

const mockedCities: ShortCity[] = [
  { id: '1', name: 'someName', temperature: 213, weatherState: WeatherState.FOG },
  { id: '2', name: 'someName2', temperature: 213, weatherState: WeatherState.MOSTLY_SUNNY },
  { id: '3', name: 'someName3', temperature: 213, weatherState: WeatherState.THUNDERSTORM },
];


test('renders all cities as expected', () => {
  // don't change for snapshot testing persistance
  const cities: ShortCity[] = [
    { id: '1', name: 'someName', temperature: 213, weatherState: WeatherState.FOG },
    { id: '2', name: 'someName2', temperature: 213, weatherState: WeatherState.MOSTLY_SUNNY },
    { id: '3', name: 'someName3', temperature: 213, weatherState: WeatherState.THUNDERSTORM },
  ];
  //
  const { toJSON } = render(
    <Provider store={createMockedStore()}>
      <CitiesList onCityPressed={() => { }} loadingCities={false} cities={cities} />
    </Provider>
  );
  expect(toJSON()).toMatchSnapshot();
});

test('shows not found text if cities list is empty and they finished loading', () => {
  const { getByTestId } = render(<CitiesList onCityPressed={() => { }} loadingCities={false} cities={[]} />);
  expect(getByTestId('CitiesList.NotFoundText')).not.toBeNull();
});


test('hides not found text if cities list is loading', () => {
  const { getByTestId } = render(
    <Provider store={createMockedStore()}>
      <CitiesList onCityPressed={() => { }} loadingCities={true} cities={[]} />
    </Provider>
  );
  expect(() => getByTestId('CitiesList.NotFoundText')).toThrowError();
});

test('hides not found text if cities list finished loading and it is not empty', () => {
  const { getByTestId } = render(
    <Provider store={createMockedStore()}>
      <CitiesList onCityPressed={() => { }} loadingCities={false} cities={mockedCities} />
    </Provider>
  );
  expect(() => getByTestId('CitiesList.NotFoundText')).toThrowError();
});

test('onCityPressed callback works as expected and cities passed correctly', () => {
  const onCityPressed = jest.fn();
  const { getAllByTestId } = render(
    <Provider store={createMockedStore()}>
      <CitiesList onCityPressed={onCityPressed} loadingCities={false} cities={mockedCities} />
    </Provider>
  );

  const citiesElements = getAllByTestId(/CitiesList.City./);
  expect(citiesElements).toHaveLength(mockedCities.length);
  citiesElements.forEach((element, index) => {
    fireEvent.press(element);
    expect(onCityPressed).toHaveBeenCalledTimes(index + 1);
    expect(onCityPressed).toHaveBeenLastCalledWith(mockedCities[index]);
  });
});
