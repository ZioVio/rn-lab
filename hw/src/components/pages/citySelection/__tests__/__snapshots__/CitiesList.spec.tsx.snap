// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all cities as expected 1`] = `
<View
  style={
    Object {
      "marginBottom": 60,
    }
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "id": "1",
          "name": "someName",
          "temperature": 213,
          "weatherState": 4,
        },
        Object {
          "id": "2",
          "name": "someName2",
          "temperature": 213,
          "weatherState": 7,
        },
        Object {
          "id": "3",
          "name": "someName3",
          "temperature": 213,
          "weatherState": 3,
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="CitiesList.City.1"
        >
          <View
            style={
              Array [
                Object {
                  "borderBottomColor": "#45464f",
                  "borderBottomWidth": 2,
                },
                Object {
                  "flexDirection": "row",
                  "paddingVertical": 15,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 3,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                size={60}
                style={
                  Array [
                    Object {
                      "color": "#fffffe",
                      "fontFamily": "PTS55F",
                    },
                    Object {
                      "fontSize": 22,
                    },
                    Object {
                      "fontSize": 60,
                    },
                    undefined,
                  ]
                }
                testID="City.Temperature"
              >
                -60°C
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 4,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#fffffe",
                      "fontFamily": "PTS55F",
                    },
                    Object {
                      "fontSize": 22,
                    },
                    undefined,
                    undefined,
                  ]
                }
              >
                someName
              </Text>
              <Text
                size={20}
                style={
                  Array [
                    Object {
                      "color": "#fffffe",
                      "fontFamily": "PTS55F",
                    },
                    Object {
                      "fontSize": 22,
                    },
                    Object {
                      "fontSize": 20,
                    },
                    Object {
                      "color": "#45464f",
                    },
                  ]
                }
                testID="City.WeatherState"
              >
                // FOG
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="CitiesList.City.2"
        >
          <View
            style={
              Array [
                Object {
                  "borderBottomColor": "#45464f",
                  "borderBottomWidth": 2,
                },
                Object {
                  "flexDirection": "row",
                  "paddingVertical": 15,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 3,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                size={60}
                style={
                  Array [
                    Object {
                      "color": "#fffffe",
                      "fontFamily": "PTS55F",
                    },
                    Object {
                      "fontSize": 22,
                    },
                    Object {
                      "fontSize": 60,
                    },
                    undefined,
                  ]
                }
                testID="City.Temperature"
              >
                -60°C
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 4,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#fffffe",
                      "fontFamily": "PTS55F",
                    },
                    Object {
                      "fontSize": 22,
                    },
                    undefined,
                    undefined,
                  ]
                }
              >
                someName2
              </Text>
              <Text
                size={20}
                style={
                  Array [
                    Object {
                      "color": "#fffffe",
                      "fontFamily": "PTS55F",
                    },
                    Object {
                      "fontSize": 22,
                    },
                    Object {
                      "fontSize": 20,
                    },
                    Object {
                      "color": "#45464f",
                    },
                  ]
                }
                testID="City.WeatherState"
              >
                // MOSTLY SUNNY
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="CitiesList.City.3"
        >
          <View
            style={
              Array [
                Object {
                  "borderBottomColor": "#45464f",
                  "borderBottomWidth": 2,
                },
                Object {
                  "flexDirection": "row",
                  "paddingVertical": 15,
                },
              ]
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 3,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                size={60}
                style={
                  Array [
                    Object {
                      "color": "#fffffe",
                      "fontFamily": "PTS55F",
                    },
                    Object {
                      "fontSize": 22,
                    },
                    Object {
                      "fontSize": 60,
                    },
                    undefined,
                  ]
                }
                testID="City.Temperature"
              >
                -60°C
              </Text>
            </View>
            <View
              style={
                Object {
                  "flex": 4,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#fffffe",
                      "fontFamily": "PTS55F",
                    },
                    Object {
                      "fontSize": 22,
                    },
                    undefined,
                    undefined,
                  ]
                }
              >
                someName3
              </Text>
              <Text
                size={20}
                style={
                  Array [
                    Object {
                      "color": "#fffffe",
                      "fontFamily": "PTS55F",
                    },
                    Object {
                      "fontSize": 22,
                    },
                    Object {
                      "fontSize": 20,
                    },
                    Object {
                      "color": "#45464f",
                    },
                  ]
                }
                testID="City.WeatherState"
              >
                // AH.. SHIT
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
